feat(Algebra/Homology): add Euler–Poincaré formula

jessealama opened the pull request:
This PR builds on the generalized Euler characteristic framework from #31121 to prove the Euler-Poincaré formula for chain complexes.

PR #31121 defines the Euler characteristic for general homological complexes. This PR specializes those definitions to ℤ-indexed chain complexes and proves the main Euler-Poincaré theorem.

### Main result (in `EulerPoincare.lean`)

* `ChainComplex.eulerChar_eq_homologyEulerChar`: For ℤ-indexed bounded chain complexes of finite-dimensional modules over a division ring, the alternating sum of chain dimensions equals the alternating sum of homology dimensions.

### Supporting lemmas

The file also provides dimension lemmas generalized to arbitrary `HomologicalComplex (ModuleCat k) c` (not just ℤ-indexed chain complexes):

* `HomologicalComplex.dFrom_zero_range` / `dTo_zero_range`: zero range when the target/source object is zero
* `HomologicalComplex.dFrom_range_finrank_eq_d` / `dTo_range_finrank_eq_d`: range of `dFrom`/`dTo` has the same dimension as the underlying differential
* `HomologicalComplex.range_dTo_le_ker_dFrom`: range of `dTo` is contained in the kernel of `dFrom`

---

- [ ] depends on: #38483 (`Antiperiodic.sum_Ico_shift` and bilinear cancellation lemma)

Builds on: #31121
Related to: #29639, #29643, #29646

kim-em commented:
(btw --- don't mark things as "draft" unless you specifically do not want review yet.)

kim-em commented on Mathlib/Algebra/Homology/EulerCharacteristic.lean:
```diff
@@ -0,0 +1,79 @@
+sum `∑_{k=0}^{n-1} (-1)^k rank(C_k)` of the ranks of the chain groups from index 0 to n-1.
+This is most commonly used when the base ring is a field (where rank means dimension)
+or a principal ideal domain. -/
+noncomputable def boundedEulerChar (C : ChainComplex (ModuleCat R) ℕ) (n : ℕ) : ℤ :=
+  ∑ k ∈ Finset.range n, (-1 : ℤ)^k * Module.finrank R (C.X k)
```
This may not be something you're interested in doing, but this could be made massively more general by defining the function from ChainComplex C Nat to the grothendieck group (whether split or exact or triangulated!) of C for any C, not just ModuleCat R.

kim-em commented on Mathlib/Algebra/Homology/EulerCharacteristic.lean:
```diff
@@ -0,0 +1,79 @@
+
+/-- The homological Euler characteristic, defined as the alternating sum
+of dimensions of homology groups up to index n. -/
+noncomputable def homologyBoundedEulerChar (C : ChainComplex (ModuleCat R) ℕ) (n : ℕ)
+    [∀ i : ℕ, (C.sc i).HasHomology] : ℤ :=
+  ∑ i ∈ Finset.range n, (-1 : ℤ)^i * Module.finrank R (C.homology i)
```
Infinite sum versions?

kim-em commented on Mathlib/Algebra/Homology/EulerPoincare.lean:
```diff
@@ -0,0 +1,319 @@
+  rw [sum_range_alternating_pow_neg_one_succ]
+  simp
+
+set_option maxHeartbeats 400000 in -- need to increase heartbeats for final simp
```
Can you `set_option diagnostics true` here and see what `simp` is doing? This seems bad.

jessealama replied:
Looks like there's quite a lot of unfolding of instances and reducible definitions, especially in `ModuleCat.carrier`. I tweaked things a bit by using a simple `rw` and then `rfl`. Phew!

kim-em commented:
(My suggestions are not necessary conditions for merging this. We can generalize later.)

jessealama commented:
> (My suggestions are not necessary conditions for merging this. We can generalize later.)

Thanks fot taking a look! I think working with Grothendieck groups might be a bit of a stretch, at this point. I think this is a decent result as it stands, though I definitely agree that an integer generalization is even better and should be possible.

jessealama commented:
> (btw --- don't mark things as "draft" unless you specifically do not want review yet.)

Sorry about that. I had pinged you about an issue related to the build failing because of the new checks on things like "nightly-testing". I didn't understand how to mitigate against the new check, which was failing but not due to any change I made here. The fix was to simply rebase; sorry for the noise!

jessealama commented on Mathlib/Algebra/Homology/EulerCharacteristic.lean:
```diff
@@ -0,0 +1,79 @@
+/-- The Euler characteristic of a chain complex, defined as an infinite sum.
+This is the alternating sum `∑_{k=0}^∞ (-1)^k rank(C_k)` of the ranks of all chain groups.
+The sum may not converge in general. -/
+noncomputable def eulerChar (C : ChainComplex (ModuleCat R) ℕ) : ℤ :=
```
Done! Care to take another look?

jessealama commented:
An earlier version of the main result was over fields, but we can do just as well with division rings, so I generalized it a bit more.

jessealama commented:
> This pull request has conflicts, please merge `master` and resolve them.

resolved!

jessealama committed:
feat(Algebra/Homology): add Euler-Poincaré formula for bounded chain complexes

Prove `ChainComplex.eulerChar_eq_homologyEulerChar`: for ℤ-indexed
bounded chain complexes of finite-dimensional modules over a division
ring, the Euler characteristic (alternating sum of dimensions) equals
the homological Euler characteristic (alternating sum of homology
dimensions).

The proof uses a telescoping sum argument: the dimension of each
chain module decomposes via rank-nullity as kernel + image, and
the kernel further decomposes as homology + boundaries.
The boundary terms telescope and cancel at adjacent degrees,
leaving only the homology terms.

This builds on the `finsum`-based Euler characteristic infrastructure
from `Mathlib.Algebra.Homology.EulerCharacteristic`.

jessealama committed:
refactor(Algebra/Homology): use Int.coe_negOnePow in Euler-Poincaré proof

jessealama committed:
refactor(Algebra/Homology): simplify Euler-Poincaré proof, eliminate duplication

- Replace hand-rolled Subsingleton proof with ModuleCat.subsingleton_of_isZero
- Delete cycles_dimension_decomposition wrapper, inline as ← homology_finrank_formula
- Golf chain_dimension_decomposition proof using omega
- Simplify Finset interval helpers to one-liners (ext; simp; omega)
- Extract isZero_outside_Ico helper to deduplicate IsZero logic in main theorem
- Generalize dimension lemmas from ChainComplex to HomologicalComplex

jessealama committed:
golf(Algebra/Homology): trim docs and golf Euler-Poincaré proof

Reduce EulerPoincare.lean from 368 to 302 lines:
- Remove/shorten docstrings on trivial and private lemmas
- Inline temporary `have` bindings in `sum_Ico_alternating_shift_decomp`
- Replace 36-line `calc` in `homology_finrank_formula` with direct proof
- Eliminate `let`/`change` pattern in main theorem via inline lambdas
- Use `rcases`/`<;>` and `exact_mod_cast` for shorter tactic blocks
- Merge consecutive `rw` calls

jessealama committed:
refactor(Algebra/Homology): extract alternating cancellation lemma, use existing Ico splits

Remove 4 custom utility lemmas from EulerPoincare.lean and replace with:
- `Finset.sum_Ico_add_sum_Ico_shift_neg_cancel` (new, in Intervals.lean)
- Existing `insert_Ico_add_one_left_eq_Ico` / `insert_Ico_right_eq_Ico_add_one`
- `Int.negOnePow_succ` + `Int.coe_negOnePow` for sign alternation

jessealama committed:
style: trim docstrings, flatten nested haves, rename section

jessealama committed:
style(Algebra/Homology): replace flexible simp with simp only in Euler-Poincaré

Replace `simp [xNext]` and `simp [xPrev]` with `simp only` followed
by explicit `rw` using `next_eq'`/`prev_eq'` to satisfy the
`linter.style.flexible` check.

jessealama committed:
refactor(Algebra/Homology): use Antiperiodic.sum_Ico_mul_add_sum_Ico_mul_shift_eq_zero

Replaces the inline `sum_Ico_add_sum_Ico_shift_neg_cancel` lemma in
`BigOperators/Intervals.lean` with a call to the new
`Antiperiodic.sum_Ico_mul_add_sum_Ico_mul_shift_eq_zero` from
`BigOperators/Periodic.lean` (added in #38483).
